qiskit/circuit/singleton_gate.py

```diff
@@ -80,6 +80,25 @@ def __init__(self, *args, _condition=None, **kwargs):
         super().__init__(*args, **kwargs)
         self._condition = _condition
 
+    def __setstate__(self, state):
+        if state is not None:
+            self.label = state[0]
+            self._condition = state[1]
+            self.duration = state[2]
+            self.unit = state[3]
```

Comment:
If the `__setstate__` part of the `__reduce__` tuple is `None`, then `__setstate__` doesn't get called at all.  I think the check has no real meaning here.

Reply:
The docs mention something like this for `__getstate__` [here](https://docs.python.org/3/library/pickle.html#object.__setstate__):

> Note: If [__getstate__()](https://docs.python.org/3/library/pickle.html#object.__getstate__) returns a false value, the [__setstate__()](https://docs.python.org/3/library/pickle.html#object.__setstate__) method will not be called upon unpickling.

but I didn't see specifically called out for `__reduce__`. I confirmed that `__setstate__` was not called for `test_immutable_pickle`.

Reply:
This isn't relevant anymore with: https://github.com/Qiskit/qiskit/commit/3ad38b4ae95f23bae69d47942162345ed71ca6dd